### PR TITLE
Add failure handling to system-agent

### PR DESCRIPTION
**Pull request by Oats87**
https://github.com/rancher/system-agent/issues/49

### CODE REVIEW

File: pkg/applyinator/applyinator.go [hunk 1]
```diff
@@ -172,7 +173,7 @@ func (a *Applyinator) Apply(ctx context.Context, cp CalculatedPlan) ([]byte, err
 	return gzOutput.Bytes(), nil
 }
 
-func (a *Applyinator) execute(ctx context.Context, executionDir string, instruction Instruction) ([]byte, error) {
+func (a *Applyinator) execute(ctx context.Context, prefix string, executionDir string, instruction Instruction) ([]byte, error) {
```
Comment (briandowns) [inline pkg/applyinator/applyinator.go:176]:
```suggestion
func (a *Applyinator) execute(ctx context.Context, prefix, executionDir string, instruction Instruction) ([]byte, error) {
```

File: pkg/applyinator/applyinator.go [hunk 2]
```diff
@@ -146,7 +146,8 @@ func (a *Applyinator) Apply(ctx context.Context, cp CalculatedPlan) ([]byte, err
 	for index, instruction := range cp.Plan.Instructions {
 		logrus.Debugf("Executing instruction %d for plan %s", index, cp.Checksum)
 		executionInstructionDir := filepath.Join(executionDir, cp.Checksum+"_"+strconv.Itoa(index))
-		output, err := a.execute(ctx, executionInstructionDir, instruction)
+		prefix := cp.Checksum + "_" + strconv.Itoa(index)
```
Comment (briandowns) [inline pkg/applyinator/applyinator.go:149]: Do we know for certain `index` will always have an integer value in it?

Comment (Oats87) [reply pkg/applyinator/applyinator.go:149]: I believe so, `cp.Plan.Instructions` is a slice

File: pkg/k8splan/watcher.go [hunk 3]
```diff
@@ -221,6 +307,27 @@ func (w *watcher) start(ctx context.Context) {
 	}
 }
 
+func (w *watcher) updateSecret(core corecontrollers.Interface, secret *v1.Secret) (*v1.Secret, error) {
+	var resultingSecret *v1.Secret
+	err := retry.OnError(retry.DefaultBackoff,
+		func(err error) bool {
+			if apierrors.IsConflict(err) {
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:314]: To be clear, we're not retrying if it's a conflict, but we will for other errors?

Comment (Oats87) [reply pkg/k8splan/watcher.go:314]: Correct. This is a request that is originally from Darren, and will handle transient errors such as GOAWAY or server unavailable etc. 

The only time we will guaranteed not be able to update is if there is a conflict as the server will never take our version, so in that case we scrap what we're trying to do

File: pkg/k8splan/watcher.go [hunk 4]
```diff
@@ -221,6 +307,27 @@ func (w *watcher) start(ctx context.Context) {
...
+				return false
+			}
+			return true
+		},
+		func() error {
+			var err error
+			resultingSecret, err = core.Secret().Update(secret)
+			return err
+		})
+	if err == nil {
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:324]: Can we refactor this so it looks for an error and actos accordily and returns nil for the resultingSecret and nil for the err at the end of the func?

File: pkg/k8splan/watcher.go [hunk 5]
```diff
@@ -153,64 +192,111 @@ func (w *watcher) start(ctx context.Context) {
...
+				}
+			} else {
+				maxFailureThreshold = int(^uint(0) >> 1) // max int
+			}
+
+			logrus.Debugf("[K8s] Parsed max failure threshold value of %d", maxFailureThreshold)
+			wasFailedPlan := false
+			if rawFailureCount, ok := secret.Data[failureCountKey]; ok {
+				failureCount, err := strconv.Atoi(string(rawFailureCount))
+				if err == nil {
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:211]: Can we reverse this logic and check for the error and just log it if it exists? We an move the logic in this block out of this indentation.

File: pkg/k8splan/watcher.go [hunk 6]
```diff
@@ -153,64 +192,111 @@ func (w *watcher) start(ctx context.Context) {
 				}
 			}
 
+			// Check to see if we've exceeded our failure count threshold
+			var maxFailureThreshold int
+			if rawMaxFailureThreshold, ok := secret.Data[maxFailuresKey]; ok {
+				// max failure threshold is defined. parse and compare
+				maxFailureThreshold, err = strconv.Atoi(string(rawMaxFailureThreshold))
+				if err != nil {
+					maxFailureThreshold = int(^uint(0) >> 1) // max int
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:201]: If no threshold is defined, is our default threshold 9,223,372,036,854,775,807?

File: pkg/k8splan/watcher.go [hunk 7]
```diff
@@ -116,12 +159,8 @@ func (w *watcher) start(ctx context.Context) {
...
 		needsApplied := true
 		if toInt(w.lastAppliedResourceVersion) > toInt(secret.ResourceVersion) {
-			logrus.Debugf("received secret to process that was older than the last secret operated on. (%s vs %s)", secret.ResourceVersion, w.lastAppliedResourceVersion)
-			return secret, nil
-		}
-		if w.lastAppliedResourceVersion == secret.ResourceVersion {
-			logrus.Debugf("last applied resource version (%s) did not change. skipping apply.", w.lastAppliedResourceVersion)
-			needsApplied = false
+			logrus.Debugf("[K8s] received secret to process that was older than the last secret operated on. (%s vs %s)", secret.ResourceVersion, w.lastAppliedResourceVersion)
+			return secret, fmt.Errorf("secret received was too old")
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:163]:
```suggestion
			return secret, errors.New("secret received was too old")
```

File: pkg/k8splan/watcher.go [hunk 8]
```diff
@@ -221,6 +307,27 @@ func (w *watcher) start(ctx context.Context) {
...
+		func() error {
+			var err error
+			resultingSecret, err = core.Secret().Update(secret)
+			return err
+		})
+	if err == nil {
+		logrus.Debugf("[K8s] updating lastAppliedResourceVersion to %s", resultingSecret.ResourceVersion)
+		w.lastAppliedResourceVersion = resultingSecret.ResourceVersion
+	}
+	return resultingSecret, err
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:328]: Is there value in returning this rather than the normal `nil` value?

Comment (Oats87) [reply pkg/k8splan/watcher.go:328]: I'm assuming you're referring to the `resultingSecret` -- we need to return the `resultingSecret` back to the controller framework, which is why the logic is all wonky here.

File: pkg/k8splan/watcher.go (hunk 4 above)
Comment (Oats87) [reply pkg/k8splan/watcher.go:324]: I am working to retain the "standard" behavior which is to always return what is resulting from `core.Secret().Update()`, but we only want to update the last applied resource version in the event that the secret update was successful i.e. where `err == nil`

File: pkg/k8splan/watcher.go (hunk 5 above)
Comment (Oats87) [reply pkg/k8splan/watcher.go:211]: I didn't want to have the rest of this code hit in the event that the conversion was unsuccessful as we don't want to cause false positives of failure detection. It should basically never be that way unless someone manually manipulates the count value in the secret or corruption occurs.

File: pkg/k8splan/watcher.go (hunk 6 above)
Comment (Oats87) [reply pkg/k8splan/watcher.go:201]: Yes. I guess I could also just set it to `-1` and have logic to handle that case below, but by setting it to max int should provide more than ample amounts of attempts of a plan in my opinion.

File: pkg/k8splan/watcher.go [hunk 9]
```diff
@@ -153,64 +193,114 @@ func (w *watcher) start(ctx context.Context) {
...
+				}
+			} else {
+				maxFailureThreshold = -1
+			}
+			wasFailedPlan := false
+			if rawFailureCount, ok := secret.Data[failureCountKey]; ok {
+				failureCount, err := strconv.Atoi(string(rawFailureCount))
+				if err != nil {
+					logrus.Errorf("[K8s] Error while parsing raw failure count: %v", err)
+					failureCount = 0
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:214]: Should we stop processing at this point?

**Pull request merged**